File: src/app/api/schemas/live_runtime.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class WorkerHeartbeatResponse(BaseModel):
    worker_id: str
    pod_name: str
    shard_id: int
    status: str
    owned_symbol_count: int
    updated_at: datetime


class ShardAssignmentResponse(BaseModel):
    shard_id: int
    symbol_keys: list[str]


class LeaseResponse(BaseModel):
    symbol_key: str
    worker_id: str
    pod_name: str
    shard_id: int
    assignment_version: int
    leased_at: datetime
    expires_at: datetime


class ControlFlagsResponse(BaseModel):
    kill_switch_enabled: bool
    paused_contracts: list[str]
    paused_symbols: list[str]
    paused_shards: list[int]


class RuntimeStateResponse(BaseModel):
    assignment_version: int
    assignments: list[ShardAssignmentResponse]
    workers: list[WorkerHeartbeatResponse]
    leases: list[LeaseResponse]
    control_flags: ControlFlagsResponse


class WorkerEventResponse(BaseModel):
    id: UUID
    worker_id: str
    shard_id: int | None
    contract_id: UUID | None
    symbol_key: str | None
    event_type: str
    severity: str
    payload: dict[str, Any]
    created_at: datetime


class LiveRuntimeResponse(BaseModel):
    state: RuntimeStateResponse
    events: list[WorkerEventResponse]


class LiveRuntimeEventsQuery(BaseModel):
    limit: int = Field(default=100, ge=1, le=500)
    worker_id: str | None = None
    event_type: str | None = None
    symbol_key: str | None = None
    since: datetime | None = None
    until: datetime | None = None

File: web/src/types/liveRuntime.ts
export interface WorkerHeartbeat {
  worker_id: string
  pod_name: string
  shard_id: number
  status: string
  owned_symbol_count: number
  updated_at: string
}

export interface ShardAssignment {
  shard_id: number
  symbol_keys: string[]
}

export interface LeaseRecord {
  symbol_key: string
  worker_id: string
  pod_name: string
  shard_id: number
  assignment_version: number
  leased_at: string
  expires_at: string
}

export interface ControlFlags {
  kill_switch_enabled: boolean
  paused_contracts: string[]
  paused_symbols: string[]
  paused_shards: number[]
}

export interface RuntimeState {
  assignment_version: number
  assignments: ShardAssignment[]
  workers: WorkerHeartbeat[]
  leases: LeaseRecord[]
  control_flags: ControlFlags
}

export type WorkerEventSeverity = 'info' | 'warn' | 'error'

export interface WorkerEvent {
  id: string
  worker_id: string
  shard_id: number | null
  contract_id: string | null
  symbol_key: string | null
  event_type: string
  severity: WorkerEventSeverity
  payload: Record<string, unknown>
  created_at: string
}

export interface LiveRuntimeResponse {
  state: RuntimeState
  events: WorkerEvent[]
}
